In [1]:
import pandas as pd

In [7]:
df = pd.read_csv('vehicles.csv', low_memory=False)

In [15]:
# range

### Question 1: Convert a text column into a categorical column. How much memory did you save? 

In [16]:
engine_description = df['eng_dscr']


In [20]:
engine_description_category = engine_description.astype('category')

In [24]:
mem_use_non_cat = engine_description.memory_usage(deep=True)
mem_use_cat = engine_description_category.memory_usage(deep=True)

In [25]:
mem_use_non_cat - mem_use_cat

1997359

Converting the column to a category saves 1,997,359 Bytes

### Question 2: Convert a numeric column into a categorical column by binning it (pd.cut). How much memory did you save?

In [27]:
car_range = df['range']

In [41]:
mem_use_non_bin = car_range.memory_usage(deep=True)

In [42]:
# car_range.groupby(car_range).size()
bin_range = [0, 50, 100, 150, 200, 250, 300, 350, 400]
car_range_bin = pd.cut(car_range, bins=bin_range, right=False)
mem_use_bin =car_range_bin.memory_usage(deep=True)

In [43]:
mem_use_non_bin - mem_use_bin

287580

Converting the column to a category saves 287,580 Bytes

### Question 3: Use the generalize_topn function to limit the amounts of categories in your column. How much memory did you save?

In [44]:
def generalize_topn(ser, n=5, other='Other'):
    topn = ser.value_counts().index[:n]
    if isinstance(ser.dtype, pd.CategoricalDtype): 
        ser = ser.cat.set_categories( topn.set_categories(list(topn)+[other])) 
    return ser.where(ser.isin(topn), other)

In [46]:
engine_description_category_limit = engine_description_category.pipe(generalize_topn, n=5, other='Other')

In [47]:
engine_description_category.memory_usage(deep=True) - engine_description_category_limit.memory_usage(deep=True)

96628

Converting the column to use top N saves 96,628 Bytes